lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,520 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import type { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { SessionContainerInterface } from "../session/types";
+import {
+    TypeInput as EmailDeliveryTypeInput,
+    TypeInputWithService as EmailDeliveryTypeInputWithService,
+} from "../../ingredients/emaildelivery/types";
+import EmailDeliveryIngredient from "../../ingredients/emaildelivery";
+import { GeneralErrorResponse, NormalisedAppinfo, User, UserContext } from "../../types";
+import RecipeUserId from "../../recipeUserId";
+
+export type TypeNormalisedInput = {
+    relyingPartyId: TypeNormalisedInputRelyingPartyId;
+    relyingPartyName: TypeNormalisedInputRelyingPartyName;
+    getOrigin: TypeNormalisedInputGetOrigin;
+    getEmailDeliveryConfig: (
+        isInServerlessEnv: boolean
+    ) => EmailDeliveryTypeInputWithService<TypeWebauthnEmailDeliveryInput>;
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInputRelyingPartyId = (input: {
+    request: BaseRequest | undefined;
+    userContext: UserContext;
+}) => string; // should return the domain of the origin
+
+export type TypeNormalisedInputRelyingPartyName = (input: {
+    request: BaseRequest | undefined;
+    userContext: UserContext;
+}) => string; // should return the app name
```

Comment:
why do these need to be functinos that take in request? Maybe instead they should take in the tenant id? What kind of use cases do you think exist that users might want to dynamically set relyingPartyId and relyingPartyName?

Reply:
Relying party id: 
If they don't pass anything, we will use the origin. 
If they pass in a string, we will use that. 
If they pass a function, they might want to make use of the request details such as origin, headers maybe, etc). 
They might want to use dynamically when running multiple tenants on different domains.

Relying party name:
i can drop the inputs and only keep the function without any params. it will default to app name.

Done